File: 2024Autumn/bigdatatest1/code1.py
import networkx as nx
from collections import defaultdict
import time

def calculate_modularity(G, communities):
    """计算给定图 G 和社区划分的模块度 Q"""
    m = G.size(weight='weight')
    Q = 0.0
    for community in communities.values():
        internal_weight = sum(G[u][v].get('weight', 1) for u in community for v in community if G.has_edge(u, v))
        degree_sum = sum(G.degree(n, weight='weight') for n in community)
        Q += internal_weight / (2 * m) - (degree_sum / (2 * m)) ** 2
    return Q

def initialize_communities(G):
    return {node: {node} for node in G.nodes()}

def modularity_gain(G, node, community, community_weight, total_weight):
    node_degree = G.degree(node, weight='weight')
    community_degree_sum = sum(G.degree(n, weight='weight') for n in community)
    links_to_community = sum(G[node][nbr].get('weight', 1) for nbr in community if G.has_edge(node, nbr))
    delta_Q = (links_to_community - (node_degree * community_degree_sum) / (2 * total_weight)) / (2 * total_weight)
    return delta_Q

def louvain_algorithm(G, target_communities=5):
    total_weight = G.size(weight='weight')
    communities = initialize_communities(G)
    node_to_community = {node: node for node in G.nodes()}

    while True:
        improved = False
        for node in G.nodes():
            original_community = node_to_community[node]
            best_community = original_community
            best_increase = 0

            communities[original_community].remove(node)
            neighbor_communities = set(node_to_community[neighbor] for neighbor in G.neighbors(node))
            for community in neighbor_communities:
                increase = modularity_gain(G, node, communities[community], total_weight, total_weight)
                if increase > best_increase:
                    best_increase = increase
                    best_community = community

            if best_community != original_community:
                node_to_community[node] = best_community
                communities[best_community].add(node)
                improved = True
            else:
                communities[original_community].add(node)

        if not improved or len(communities) <= target_communities:
            break

        new_G = nx.Graph()
        new_node_to_community = {}
        for community, nodes in communities.items():
            new_node = community
            new_G.add_node(new_node)
            new_node_to_community[new_node] = nodes
            for node in nodes:
                for neighbor in G.neighbors(node):
                    neighbor_community = node_to_community[neighbor]
                    weight = G[node][neighbor].get('weight', 1)
                    if new_G.has_edge(new_node, neighbor_community):
                        new_G[new_node][neighbor_community]['weight'] += weight
                    else:
                        new_G.add_edge(new_node, neighbor_community, weight=weight)

        G = new_G
        communities = new_node_to_community
        node_to_community = {node: community for community, nodes in communities.items() for node in nodes}

    return {community: nodes for community, nodes in communities.items() if nodes}

# 初始化无向图
G = nx.Graph()

# 读取文件并加载图结构
with open("Cit-HepPh.txt", "r") as file:
    for line in file:
        if line.startswith("#"):
            continue
        from_node, to_node = map(int, line.strip().split())
        G.add_edge(from_node, to_node)

# 获取图的信息
graph_info = f"图的节点数: {G.number_of_nodes()}\n图的边数: {G.number_of_edges()}\n"

# 记录程序开始时间
start_time = time.time()

# 应用 Louvain 算法进行社区划分，指定目标社区数为 5
communities = louvain_algorithm(G, target_communities=5)

# 记录程序结束时间
end_time = time.time()

# 计算程序运行时间
execution_time = f"程序运行时间: {end_time - start_time:.2f} 秒\n"

# 准备要写入的内容
result_content = graph_info + "最终社区划分：\n"
for community, nodes in communities.items():
    result_content += f"社区 {community}: 包含节点 {nodes}\n"

result_content += execution_time

# 将结果写入文件 result.txt
with open("code1ans.txt", "w") as result_file:
    result_file.write(result_content)

print("结果已保存到 code1ans.txt 文件中")

File: 2024Autumn/bigdatatest1/code1.scala
import scala.collection.mutable
import scala.io.Source
import System.nanoTime

object LouvainAlgorithm {

  case class Graph(edges: mutable.Map[Int, mutable.Map[Int, Double]], nodes: mutable.Set[Int]) {
    def addEdge(from: Int, to: Int, weight: Double): Unit = {
      if (!edges.contains(from)) edges(from) = mutable.Map()
      if (!edges.contains(to)) edges(to) = mutable.Map()

      edges(from)(to) = weight
      edges(to)(from) = weight
    }

    def getEdgeWeight(from: Int, to: Int): Double = {
      edges.getOrElse(from, mutable.Map()).getOrElse(to, 1.0) // Default weight 1 for unweighted graphs
    }

    def size(): Double = {
      edges.values.flatMap(_.values).sum / 2
    }

    def degree(node: Int): Double = {
      edges.getOrElse(node, mutable.Map()).values.sum
    }

    def neighbors(node: Int): Set[Int] = {
      edges.getOrElse(node, mutable.Map()).keys.toSet
    }
  }

  // 计算模块度
  def calculateModularity(graph: Graph, communities: Map[Int, Set[Int]]): Double = {
    val m = graph.size()
    var Q = 0.0

    for ((community, nodes) <- communities) {
      val internalWeight = nodes.flatMap(u => nodes.map(v => if (u != v) graph.getEdgeWeight(u, v) else 0.0)).sum
      val degreeSum = nodes.map(graph.degree).sum
      Q += internalWeight / (2 * m) - Math.pow(degreeSum / (2 * m), 2)
    }

    Q
  }

  // 初始化每个节点为一个独立的社区
  def initializeCommunities(graph: Graph): Map[Int, Set[Int]] = {
    graph.nodes.map(node => node -> Set(node)).toMap
  }

  // 计算模块度增益
  def modularityGain(graph: Graph, node: Int, community: Set[Int], communityWeight: Double, totalWeight: Double): Double = {
    val nodeDegree = graph.degree(node)
    val communityDegreeSum = community.map(graph.degree).sum
    val linksToCommunity = community.map(n => graph.getEdgeWeight(node, n)).sum
    (linksToCommunity - (nodeDegree * communityDegreeSum) / (2 * totalWeight)) / (2 * totalWeight)
  }

  // Louvain算法
  def louvainAlgorithm(graph: Graph, targetCommunities: Int): Map[Int, Set[Int]] = {
    var communities = initializeCommunities(graph)
    var nodeToCommunity = graph.nodes.map(node => node -> node).toMap

    var improved = true
    while (improved) {
      improved = false

      // 节点遍历，计算增益并移动到最佳社区
      for (node <- graph.nodes) {
        val originalCommunity = nodeToCommunity(node)
        var bestCommunity = originalCommunity
        var bestIncrease = 0.0

        communities = communities.updated(originalCommunity, communities(originalCommunity) - node)

        // 获取相邻社区
        val neighborCommunities = graph.neighbors(node).map(nodeToCommunity).toSet

        for (community <- neighborCommunities) {
          val increase = modularityGain(graph, node, communities(community), graph.size(), graph.size())
          if (increase > bestIncrease) {
            bestIncrease = increase
            bestCommunity = community
          }
        }

        if (bestCommunity != originalCommunity) {
          nodeToCommunity = nodeToCommunity.updated(node, bestCommunity)
          communities = communities.updated(bestCommunity, communities(bestCommunity) + node)
          improved = true
        } else {
          communities = communities.updated(originalCommunity, communities(originalCommunity) + node)
        }
      }

      // 如果没有改进或达到目标社区数，结束
      if (!improved || communities.size <= targetCommunities) {
        break
      }

      // 重建图，并进行下一轮的社区发现
      var newGraph = new Graph(mutable.Map(), mutable.Set())
      var newCommunities = mutable.Map[Int, Set[Int]]()
      for ((community, nodes) <- communities) {
        val newNode = community
        newGraph.nodes += newNode
        newCommunities(newNode) = nodes
        for (node <- nodes) {
          for (neighbor <- graph.neighbors(node)) {
            val neighborCommunity = nodeToCommunity(neighbor)
            val weight = graph.getEdgeWeight(node, neighbor)
            newGraph.addEdge(newNode, neighborCommunity, weight)
          }
        }
      }

      graph = newGraph
      communities = newCommunities.toMap
      nodeToCommunity = communities.flatMap { case (community, nodes) => nodes.map(node => node -> community) }
    }

    communities
  }

  def main(args: Array[String]): Unit = {
    // 初始化图
    val graph = new Graph(mutable.Map(), mutable.Set())

    // 读取文件并加载图结构
    for (line <- Source.fromFile("Cit-HepPh.txt").getLines()) {
      if (!line.startsWith("#")) {
        val Array(fromNode, toNode) = line.split("\\s+").map(_.toInt)
        graph.addEdge(fromNode, toNode, 1.0)
      }
    }

    // 获取图的信息
    println(s"图的节点数: ${graph.nodes.size}")
    println(s"图的边数: ${graph.edges.size}")

    // 记录程序开始时间
    val startTime = nanoTime()

    // 应用 Louvain 算法进行社区划分，指定目标社区数为 5
    val communities = louvainAlgorithm(graph, targetCommunities = 5)

    // 记录程序结束时间
    val endTime = nanoTime()

    // 计算程序运行时间
    val executionTime = (endTime - startTime) / 1e9
    println(f"程序运行时间: $executionTime%.2f 秒")

    // 输出结果
    println("最终社区划分：")
    for ((community, nodes) <- communities) {
      println(s"社区 $community: 包含节点 ${nodes.mkString(", ")}")
    }

    // 将结果写入文件
    val resultContent = new StringBuilder
    resultContent.append(s"图的节点数: ${graph.nodes.size}\n")
    resultContent.append(s"图的边数: ${graph.edges.size}\n")
    resultContent.append("最终社区划分：\n")
    for ((community, nodes) <- communities) {
      resultContent.append(s"社区 $community: 包含节点 ${nodes.mkString(", ")}\n")
    }
    resultContent.append(f"程序运行时间: $executionTime%.2f 秒\n")

    import java.io._
    val pw = new PrintWriter(new File("code1ans.txt"))
    pw.write(resultContent.toString)
    pw.close()

    println("结果已保存到 code1ans.txt 文件中")
  }
}
